src/UniswapV4DeployerCompetition.sol

```diff
@@ -0,0 +1,87 @@
+pragma solidity 0.8.26;
+
+import {Create2} from "@openzeppelin/contracts/utils/Create2.sol";
+import {Owned} from "solmate/src/auth/Owned.sol";
+import {VanityAddressLib} from "./libraries/VanityAddressLib.sol";
+import {IUniswapV4DeployerCompetition} from "./interfaces/IUniswapV4DeployerCompetition.sol";
+
+/// @title UniswapV4DeployerCompetition
+/// @notice A contract to crowdsource a salt for the best Uniswap V4 address
+contract UniswapV4DeployerCompetition is IUniswapV4DeployerCompetition {
+    using VanityAddressLib for address;
+
+    /// @dev The salt for the best address found so far
+    bytes32 public bestAddressSalt;
+    /// @dev The best address found so far
+    address public bestAddress;
+    /// @dev The submitter of the best address found so far
+    address public bestAddressSubmitter;
+
+    /// @dev The owner of the v4 contract
+    address public immutable v4Owner;
+    /// @dev The deadline for the competition
+    uint256 public immutable competitionDeadline;
+    /// @dev The init code hash of the V4 contract
+    bytes32 public immutable initCodeHash;
+
+    /// @dev The deployer who can initiate the deployment of the v4 PoolManager, until the exclusive deploy deadline.
+    /// @dev After this deadline anyone can deploy.
+    address public immutable deployer;
+    /// @dev The deadline for exclusive deployment by deployer after deadline
+    uint256 public immutable exclusiveDeployDeadline;
+
+    constructor(
+        bytes32 _initCodeHash,
+        address _v4Owner,
+        uint256 _competitionDeadline,
+        address _exclusiveDeployer,
+        uint256 _exclusiveDeployLength
+    ) {
+        initCodeHash = _initCodeHash;
+        v4Owner = _v4Owner;
+        competitionDeadline = _competitionDeadline;
+        exclusiveDeployDeadline = _competitionDeadline + _exclusiveDeployLength;
+        deployer = _exclusiveDeployer;
+    }
+
+    /// @inheritdoc IUniswapV4DeployerCompetition
+    function updateBestAddress(bytes32 salt) external override {
+        if (block.timestamp > competitionDeadline) {
+            revert CompetitionOver(block.timestamp, competitionDeadline);
+        }
+
+        address saltSubAddress = address(bytes20(salt));
+        if (saltSubAddress != msg.sender && saltSubAddress != address(0)) revert InvalidSender(salt, msg.sender);
+
+        address newAddress = Create2.computeAddress(salt, initCodeHash, address(this));
+        if (bestAddress != address(0) && !newAddress.betterThan(bestAddress)) {
+            revert WorseAddress(newAddress, bestAddress, newAddress.score(), bestAddress.score());
+        }
+
+        bestAddress = newAddress;
+        bestAddressSalt = salt;
+        bestAddressSubmitter = msg.sender;
+
+        emit NewAddressFound(newAddress, msg.sender, newAddress.score());
+    }
+
+    /// @inheritdoc IUniswapV4DeployerCompetition
+    function deploy(bytes memory bytecode) external override {
+        if (keccak256(bytecode) != initCodeHash) {
```

Comment:
GAS

Since only a certain bytecode is expected to be deployed, it can be stored in this contract's code. So instead of taking it as an argument, it can be deployed like `Create2.deploy(type(ContractToDeploy).creationCode)`. This will save gas on storing `initCodeHash` and taking `bytecode` as an argument, but will increase the cost of the deployment of this contract, so it should be checked additionally.

Reply:
Might be not that beneficial since since it's deployed only once.

Reply:
I think its still a question of why require them to pass `bytecode` why not just use `initCodeHash` and save this code and custom error?

Reply:
want constructor params to be included as well, and think this approach is more clear for salt grinders since they can easily fetch the initcodehash they need rather than having to calculate it manually. not too stressed about minor gas for single use